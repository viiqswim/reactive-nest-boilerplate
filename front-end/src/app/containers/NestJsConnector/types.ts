import { User } from 'types/Repo';

/* --- STATE --- */
export interface NestJsConnectorState {
  userId: number;
  loading: boolean;
  error?: RepoErrorType | null;
  user: User;
}

export enum RepoErrorType {
  RESPONSE_ERROR = 1,
  USER_NOT_FOUND = 2,
  USER_ID_EMPTY = 3,
  USER_HAS_NO_USER = 4,
}

/*
  If you want to use 'ContainerState' keyword everywhere in your feature folder,
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = NestJsConnectorState;
