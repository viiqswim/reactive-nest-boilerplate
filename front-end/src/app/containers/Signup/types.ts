import { User } from 'types/User';

/* --- STATE --- */
export interface SignupState {
  id?: string;
  isLoggedIn?: boolean;
  userEmail?: string;
  password?: string;
  passwordConfirm?: string;
  loading?: boolean;
  error?: UserErrorType | null;
  user?: User;
}

export enum UserErrorType {
  RESPONSE_ERROR = 1,
  USER_NOT_FOUND = 2,
  PASSWORDS_DO_NOT_MATCH = 3,
  USER_HAS_NO_USER = 4,
}

/*
  If you want to use 'ContainerState' keyword everywhere in your feature folder,
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = SignupState;
