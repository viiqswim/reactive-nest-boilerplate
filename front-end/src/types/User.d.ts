export interface User {
  id?: number;
  uid?: string;
  firstName?: string;
  userEmail?: string | null;
  password?: string;
}
