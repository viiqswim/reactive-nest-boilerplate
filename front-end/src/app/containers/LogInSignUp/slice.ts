/*
 * LogInSignUp Slice
 *
 * Here we define:
 * - The shape of our container's slice of Redux store,
 * - All the actions which can be triggered for this slice, including their effects on the store.
 *
 * Note that, while we are using dot notation in our reducer, we are not actually mutating the state
 * manually. Under the hood, we use immer to apply these updates to a new copy of the state.
 * Please see https://immerjs.github.io/immer/docs/introduction for more information.
 *
 */

import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState, UserErrorType } from './types';
import { User } from 'types/User';

// The initial state of the LogInSignUp container
export const initialState: ContainerState = {
  isLoggedIn: false,
  userEmail: '',
  password: '',
  user: {},
  loading: false,
  error: null,
};

const logInSignUpSlice = createSlice({
  name: 'logInSignUp',
  initialState,
  reducers: {
    changeIsLoggedIn(state, action: PayloadAction<any>) {
      state.isLoggedIn = action.payload;
    },
    changeUserEmail(state, action: PayloadAction<any>) {
      state.userEmail = action.payload;
    },
    changePassword(state, action: PayloadAction<any>) {
      state.password = action.payload;
    },
    loadUser(state) {
      state.loading = true;
      state.error = null;
      state.user = {};
    },
    userLoaded(state, action: PayloadAction<User>) {
      const user = action.payload;
      state.user = {
        userEmail: user.userEmail,
        id: 1,
        uid: user.uid,
      };
      state.loading = false;
    },
    userError(state, action: PayloadAction<UserErrorType>) {
      state.error = action.payload;
      state.loading = false;
    },
    loginUser(state) {
      state.loading = true;
      state.error = null;
      state.user = {};
    },
  },
});

export const { actions, reducer, name: sliceKey } = logInSignUpSlice;
