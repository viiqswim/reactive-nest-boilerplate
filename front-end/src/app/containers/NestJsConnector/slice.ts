/*
 * NestJsConnector Slice
 *
 * Here we define:
 * - The shape of our container's slice of Redux store,
 * - All the actions which can be triggered for this slice, including their effects on the store.
 *
 * Note that, while we are using dot notation in our reducer, we are not actually mutating the state
 * manually. Under the hood, we use immer to apply these updates to a new copy of the state.
 * Please see https://immerjs.github.io/immer/docs/introduction for more information.
 *
 */

import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState, UserErrorType } from './types';
import { User } from 'types/User';

// The initial state of the NestJsConnector container
export const initialState: ContainerState = {
  userId: 22,
  user: {},
  loading: false,
  error: null,
};

const nestJsConnectorSlice = createSlice({
  name: 'nestJsConnector',
  initialState,
  reducers: {
    changeUserId(state, action: PayloadAction<number>) {
      state.userId = Number(action.payload);
    },
    loadUser(state) {
      state.loading = true;
      state.error = null;
      state.user = {};
    },
    userLoaded(state, action: PayloadAction<User>) {
      const user = action.payload;
      state.user = user;
      state.loading = false;
    },
    userError(state, action: PayloadAction<UserErrorType>) {
      state.error = action.payload;
      state.loading = false;
    },
  },
});

export const { actions, reducer, name: sliceKey } = nestJsConnectorSlice;
