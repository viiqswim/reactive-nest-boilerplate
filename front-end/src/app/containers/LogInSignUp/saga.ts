import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { selectUserEmail, selectPassword, selectIsLoggedIn } from './selectors';
import { actions } from './slice';
import { User } from 'types/User';
import { UserErrorType } from './types';
import { firebaseApp } from '../../../firebaseApp';

export function* loginUser() {
  yield delay(500);
  // Select userEmail from store
  const userEmail: string = yield select(selectUserEmail);
  const password: string = yield select(selectPassword);

  if (userEmail.length === 0) {
    yield put(actions.userError(UserErrorType.USER_ID_EMPTY));
    return;
  }

  try {
    const user: User = yield call(login, userEmail, password);
    if (user.uid) {
      yield put(actions.userLoaded(user));
    } else {
      yield put(actions.userError(UserErrorType.USER_HAS_NO_USER));
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.userError(UserErrorType.USER_NOT_FOUND));
    } else {
      yield put(actions.userError(UserErrorType.RESPONSE_ERROR));
    }
  }
}

export function* logoutUser() {
  yield delay(500);
  const isLoggedIn: boolean = yield select(selectIsLoggedIn);
  if (!isLoggedIn) {
    return;
  }

  try {
    const user = yield call(logout);
    if (user.uid) {
      yield put(actions.changeIsLoggedIn(false));
    } else {
      yield put(actions.userError(UserErrorType.RESPONSE_ERROR));
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.userError(UserErrorType.USER_NOT_FOUND));
    } else {
      yield put(actions.userError(UserErrorType.RESPONSE_ERROR));
    }
  }
}

export async function login(userEmail: string, password: string): Promise<any> {
  const response = await firebaseApp
    .auth()
    .signInWithEmailAndPassword(userEmail, password);

  return {
    uid: response?.user?.uid,
    userEmail: response?.user?.email,
  };
}

export async function logout(): Promise<any> {
  const response = await firebaseApp
    .auth()
    .signOut()
    .catch(function (error) {
      var errorCode = error.code;
      var errorMessage = error.message;
      console.log(errorCode);
      console.log(errorMessage);
    });

  return response;
}

/**
 * Root saga manages watcher lifecycle
 */
export function* logInSignUpSaga() {
  // Watches for loadUser actions and calls getUser when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loginUser.type, loginUser);
  yield takeLatest(actions.logoutUser.type, logoutUser);
}
