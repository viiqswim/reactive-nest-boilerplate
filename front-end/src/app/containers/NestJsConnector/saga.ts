import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { request } from 'utils/request';
import { selectUsername } from './selectors';
import { actions } from './slice';
import { User } from 'types/Repo';
import { RepoErrorType } from './types';

/**
 * Github repos request/response handler
 */
export function* getRepos() {
  yield delay(500);
  // Select username from store
  const username: string = yield select(selectUsername);
  if (username.length === 0) {
    yield put(actions.repoError(RepoErrorType.USERNAME_EMPTY));
    return;
  }
  const requestURL = `http://localhost:3001/users/${username}`;

  try {
    // Call our request helper (see 'utils/request')
    const user: User = yield call(request, requestURL);
    if (user.id) {
      yield put(actions.userLoaded(user));
    } else {
      yield put(actions.repoError(RepoErrorType.USER_HAS_NO_REPO));
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.repoError(RepoErrorType.USER_NOT_FOUND));
    } else {
      yield put(actions.repoError(RepoErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* nestJsConnectorSaga() {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loadRepos.type, getRepos);
}
