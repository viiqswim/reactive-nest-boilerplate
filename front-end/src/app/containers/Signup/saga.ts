import { call, put, select, takeLatest } from 'redux-saga/effects';
import { selectUserEmail, selectPassword, selectPasswordConfirm, selectIsLoggedIn } from './selectors';
import { actions } from './slice';
import { User } from 'types/User';
import { UserErrorType } from './types';
import { firebaseApp } from '../../../firebaseApp';

export function* signupUser() {
  // Select userEmail from store
  const userEmail: string = yield select(selectUserEmail);
  const password: string = yield select(selectPassword);
  const passwordConfirm: string = yield select(selectPasswordConfirm);

  if (password !== passwordConfirm) {
    yield put(actions.userError(UserErrorType.PASSWORDS_DO_NOT_MATCH));
    return;
  }

  try {
    const user: User = yield call(signup, userEmail, password);
    if (user.uid) {
      yield put(actions.userLoaded(user));
    } else {
      yield put(actions.userError(UserErrorType.USER_HAS_NO_USER));
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.userError(UserErrorType.USER_NOT_FOUND));
    } else {
      yield put(actions.userError(UserErrorType.RESPONSE_ERROR));
    }
  }
}

export async function signup(
  userEmail: string,
  password: string,
): Promise<any> {
  const response = await firebaseApp
    .auth()
    .createUserWithEmailAndPassword(userEmail, password);

  return {
    uid: response?.user?.uid,
    userEmail: response?.user?.email,
  };
}

/**
 * Root saga manages watcher lifecycle
 */
export function* signupSaga() {
  // Watches for loadUser actions and calls getUser when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.signupUser.type, signupUser);
}
