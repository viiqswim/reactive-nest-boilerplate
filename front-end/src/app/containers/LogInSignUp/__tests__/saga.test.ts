import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '../slice';

import { logInSignUpSaga, getUser } from '../saga';
import { UserErrorType } from '../types';
import { User } from 'types/User';

describe('getUser Saga', () => {
  let userId: any;
  let user: User;
  let getUserIterator: any;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getUserIterator = getUser();
    const delayDescriptor = getUserIterator.next().value;
    expect(delayDescriptor).toMatchSnapshot();

    const selectDescriptor = getUserIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should return error if userId is empty', () => {
    userId = '';
    const putDescriptor = getUserIterator.next(userId).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.userError(UserErrorType.USER_ID_EMPTY)),
    );

    const iteration = getUserIterator.next();
    expect(iteration.done).toBe(true);
  });

  it('should dispatch the userLoaded action if it requests the data successfully', () => {
    userId = 'test';
    user = {
      id: 1,
      firstName: 'user1',
    };

    const requestDescriptor = getUserIterator.next(userId).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUserIterator.next(user).value;
    expect(putDescriptor).toEqual(put(slice.actions.userLoaded(user)));
  });

  it('should dispatch the user not found error', () => {
    userId = 'test';

    const requestDescriptor = getUserIterator.next(userId).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUserIterator.throw({ response: { status: 404 } })
      .value;
    expect(putDescriptor).toEqual(
      put(slice.actions.userError(UserErrorType.USER_NOT_FOUND)),
    );
  });
  it('should dispatch the user has no user error', () => {
    userId = 'test';
    user = {};

    const requestDescriptor = getUserIterator.next(userId).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUserIterator.next(user).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.userError(UserErrorType.USER_HAS_NO_USER)),
    );
  });

  it('should dispatch the response error', () => {
    userId = 'test';

    const requestDescriptor = getUserIterator.next(userId).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getUserIterator.throw(new Error('some error')).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.userError(UserErrorType.RESPONSE_ERROR)),
    );
  });
});

describe('logInSignUpSaga Saga', () => {
  const nestJsConnectorIterator = logInSignUpSaga();
  it('should start task to watch for loadUser action', () => {
    const takeLatestDescriptor = nestJsConnectorIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.actions.loadUser.type, getUser),
    );
  });
});
