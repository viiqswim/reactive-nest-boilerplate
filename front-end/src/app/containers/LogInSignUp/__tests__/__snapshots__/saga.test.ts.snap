// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUser Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getUser Saga should dispatch the response error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/users/test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the user has no user error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the user has no user error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getUser Saga should dispatch the user has no user error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/users/test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the user not found error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the user not found error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getUser Saga should dispatch the user not found error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/users/test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the userLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the userLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getUser Saga should dispatch the userLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "undefined/users/test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should return error if userId is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should return error if userId is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
